File: ios/MemberApp/Auth/AuthManager.swift
import AppAuth
import Foundation
import Observation
import UIKit

/// OIDC authorization-code + PKCE against the STSA authentik tenant.
///
/// Design is driven by the 5-minute access-token lifetime: nothing ever holds
/// an access token string. `accessToken()` goes through
/// `OIDAuthState.performAction(freshTokens:)` on every call, which returns the
/// cached token if it is still valid and silently redeems the refresh token if
/// it is not. Callers get a token that is good *right now* or an error.
@MainActor
@Observable
final class AuthManager {
    enum AuthError: LocalizedError {
        case notAuthenticated
        case sessionExpired
        case noPresenter
        case missingUserinfoEndpoint
        case userinfoFailed(status: Int, body: String)
        case tokenUnavailable

        var errorDescription: String? {
            switch self {
            case .notAuthenticated:
                "Not signed in."
            case .sessionExpired:
                String(localized: "登入已失效，請重新登入。")
            case .noPresenter:
                "No foreground window to present the sign-in page from."
            case .missingUserinfoEndpoint:
                "The discovery document did not advertise a userinfo endpoint."
            case .userinfoFailed(let status, let body):
                "userinfo returned HTTP \(status): \(body)"
            case .tokenUnavailable:
                "The authorization server did not return an access token."
            }
        }
    }

    private static let keychainService = "tw.stsa.membership.oidc"
    private static let keychainAccount = "authState"
    private static let subjectDefaultsKey = "auth.currentSubject"

    private(set) var isLoggedIn = false
    private(set) var profile: Profile?
    private(set) var isBusy = false

    @ObservationIgnored private var authState: OIDAuthState?
    @ObservationIgnored private var userAgentSession: (any OIDExternalUserAgentSession)?
    @ObservationIgnored private var serviceConfiguration: OIDServiceConfiguration?

    init() {
        restore()
    }

    // MARK: - Session lifecycle

    func login() async throws {
        isBusy = true
        defer { isBusy = false }

        let configuration = try await discoverConfiguration()

        // The standard initialiser (no clientSecret overload) derives a PKCE
        // code_verifier and an S256 code_challenge on its own. Do not replace
        // it with the clientSecret variant — this client is public.
        let request = OIDAuthorizationRequest(
            configuration: configuration,
            clientId: AuthConfiguration.clientID,
            scopes: AuthConfiguration.scopes,
            redirectURL: AuthConfiguration.redirectURI,
            responseType: OIDResponseTypeCode,
            additionalParameters: nil
        )

        let presenter = try topViewController()

        // AppAuth drives ASWebAuthenticationSession here, so the sign-in page
        // runs in Safari's session and participates in SSO. Never swap this for
        // a WKWebView.
        let state: OIDAuthState = try await withCheckedThrowingContinuation { continuation in
            userAgentSession = OIDAuthState.authState(
                byPresenting: request,
                presenting: presenter
            ) { authState, error in
                if let authState {
                    continuation.resume(returning: authState)
                } else {
                    continuation.resume(throwing: error ?? AuthError.tokenUnavailable)
                }
            }
        }

        userAgentSession = nil
        adopt(state)

        let (profile, rawUserinfo) = try await fetchProfile()
        self.profile = profile
        cache(profile)

        logLoginResult(rawUserinfo: rawUserinfo, state: state)
    }

    /// Drops every local credential.
    ///
    /// Deliberately local-only. Because the flow runs in the shared Safari
    /// session, the authentik session cookie survives — the next `login()` may
    /// complete without a prompt. If you need to end the IdP session too, that
    /// is an RP-initiated logout against
    /// `serviceConfiguration?.discoveryDocument?.endSessionEndpoint`.
    func logout() {
        if let sub = profile?.sub ?? UserDefaults.standard.string(forKey: Self.subjectDefaultsKey) {
            UserDefaults.standard.removeObject(forKey: Self.profileKey(for: sub))
        }
        UserDefaults.standard.removeObject(forKey: Self.subjectDefaultsKey)
        try? Keychain.remove(service: Self.keychainService, account: Self.keychainAccount)

        userAgentSession = nil
        authState = nil
        profile = nil
        isLoggedIn = false
    }

    /// True when the person dismissed the sign-in sheet themselves. Callers
    /// should treat this as "nothing happened", not as a failure worth an alert.
    static func isUserCancellation(_ error: any Error) -> Bool {
        let error = error as NSError
        return error.domain == OIDGeneralErrorDomain
            && error.code == OIDErrorCode.userCanceledAuthorizationFlow.rawValue
    }

    /// Hands the redirect back to the in-flight authorization request.
    /// Wired up from `.onOpenURL` in `MemberAppApp`.
    @discardableResult
    func resume(_ url: URL) -> Bool {
        guard let session = userAgentSession else { return false }
        let handled = session.resumeExternalUserAgentFlow(with: url)
        if handled { userAgentSession = nil }
        return handled
    }

    // MARK: - Tokens

    /// The only supported way to get an access token.
    ///
    /// Never store the return value. Access tokens live 5 minutes; call this
    /// again for the next request and let AppAuth decide whether a refresh is
    /// needed.
    func accessToken() async throws -> String {
        guard let authState else { throw AuthError.notAuthenticated }

        do {
            let token: String = try await withCheckedThrowingContinuation { continuation in
                authState.performAction { accessToken, _, error in
                    if let error {
                        continuation.resume(throwing: error)
                    } else if let accessToken {
                        continuation.resume(returning: accessToken)
                    } else {
                        continuation.resume(throwing: AuthError.tokenUnavailable)
                    }
                }
            }

            // A refresh rotates the token response (and on authentik, often the
            // refresh token too), so write the updated state back.
            persist(authState)
            isLoggedIn = authState.isAuthorized

            return token
        } catch {
            // A session dying is ordinary: authentik rotates refresh tokens, they
            // expire after 30 days, and they can be revoked from the account page.
            // What is not ordinary is what used to happen next — this method threw
            // before it could update `isLoggedIn`, so the app stayed on the
            // signed-in screen with a session that could never work again, showing
            // a raw OAuth error and a Retry button that could not succeed. The
            // only way out was to find 登出 in 設定.
            //
            // AppAuth has already folded a permanent OAuth failure into the state
            // (`updateWithTokenResponse` routes `OIDOAuthTokenErrorDomain` to
            // `updateWithAuthorizationError`), so it can be asked.
            if authState.authorizationError != nil {
                logout()
                throw AuthError.sessionExpired
            }

            // Transient failure — a rotation may still have landed before it, so
            // keep whatever AppAuth wrote rather than leaving a superseded refresh
            // token in the keychain.
            persist(authState)
            throw error
        }
    }

    /// Renews the access token if it has lapsed, so the next request does not
    /// pay for a token exchange first.
    ///
    /// Called when the app comes to the foreground — not on a timer. An access
    /// token is only needed at the moment of a request, and `performAction`
    /// already renews one transparently, so polling would spend battery and
    /// rotate refresh tokens for nothing. Safe to call often: AppAuth returns
    /// the cached token while it is fresh and queues concurrent callers onto a
    /// single refresh when it is not.
    func refreshIfNeeded() async {
        guard isLoggedIn else { return }
        _ = try? await accessToken()

        // The keychain survives app deletion but UserDefaults does not, so a
        // reinstall restores the session with no cached profile — signed in,
        // but with no idea who the member is. Re-fetch rather than showing an
        // account screen with blank fields.
        if profile == nil, let (fetched, _) = try? await fetchProfile() {
            profile = fetched
            cache(fetched)
        }
    }

    /// Builds a request carrying a token that is fresh at the moment of the call.
    /// Use this for every outbound API call rather than holding a token.
    func authorizedRequest(for url: URL) async throws -> URLRequest {
        var request = URLRequest(url: url)
        request.setValue("Bearer \(try await accessToken())", forHTTPHeaderField: "Authorization")
        return request
    }

    // MARK: - Discovery

    private func discoverConfiguration() async throws -> OIDServiceConfiguration {
        if let serviceConfiguration { return serviceConfiguration }

        let configuration: OIDServiceConfiguration = try await withCheckedThrowingContinuation { continuation in
            OIDAuthorizationService.discoverConfiguration(forIssuer: AuthConfiguration.issuer) { configuration, error in
                if let configuration {
                    continuation.resume(returning: configuration)
                } else {
                    continuation.resume(throwing: error ?? AuthError.tokenUnavailable)
                }
            }
        }

        serviceConfiguration = configuration
        return configuration
    }

    // MARK: - Userinfo

    /// Returns the decoded profile plus the raw body, so the caller can log the
    /// exact claim set authentik issued.
    private func fetchProfile() async throws -> (Profile, Data) {
        let configuration = try await discoverConfiguration()
        guard let endpoint = configuration.discoveryDocument?.userinfoEndpoint else {
            throw AuthError.missingUserinfoEndpoint
        }

        let request = try await authorizedRequest(for: endpoint)
        let (data, response) = try await URLSession.shared.data(for: request)

        if let http = response as? HTTPURLResponse, http.statusCode != 200 {
            throw AuthError.userinfoFailed(
                status: http.statusCode,
                body: String(data: data, encoding: .utf8) ?? "<binary>"
            )
        }

        return (try JSONDecoder().decode(Profile.self, from: data), data)
    }

    // MARK: - Persistence

    private func adopt(_ state: OIDAuthState) {
        authState = state
        isLoggedIn = state.isAuthorized
        persist(state)
    }

    private func persist(_ state: OIDAuthState) {
        do {
            let data = try NSKeyedArchiver.archivedData(withRootObject: state, requiringSecureCoding: true)
            try Keychain.set(data, service: Self.keychainService, account: Self.keychainAccount)
        } catch {
            // Losing the write means the user re-authenticates next launch —
            // bad UX, not a correctness problem. Never fall back to a file.
            print("[Auth] Could not persist auth state to keychain: \(error)")
        }
    }

    private func restore() {
        guard let data = try? Keychain.get(service: Self.keychainService, account: Self.keychainAccount),
              let state = try? NSKeyedUnarchiver.unarchivedObject(ofClass: OIDAuthState.self, from: data)
        else { return }

        authState = state
        isLoggedIn = state.isAuthorized

        if let sub = UserDefaults.standard.string(forKey: Self.subjectDefaultsKey),
           let cached = UserDefaults.standard.data(forKey: Self.profileKey(for: sub)) {
            profile = try? JSONDecoder().decode(Profile.self, from: cached)
        }
    }

    /// Profile claims are not credentials, so they live in UserDefaults — but
    /// keyed on `sub`, never on email or username.
    private func cache(_ profile: Profile) {
        guard let data = try? JSONEncoder().encode(profile) else { return }
        UserDefaults.standard.set(data, forKey: Self.profileKey(for: profile.sub))
        UserDefaults.standard.set(profile.sub, forKey: Self.subjectDefaultsKey)
    }

    private static func profileKey(for sub: String) -> String { "auth.profile.\(sub)" }

    // MARK: - Verification logging

    private func logLoginResult(rawUserinfo: Data, state: OIDAuthState) {
        #if DEBUG
        let pretty: String = {
            guard let object = try? JSONSerialization.jsonObject(with: rawUserinfo),
                  let data = try? JSONSerialization.data(withJSONObject: object, options: [.prettyPrinted, .sortedKeys]),
                  let string = String(data: data, encoding: .utf8)
            else { return String(data: rawUserinfo, encoding: .utf8) ?? "<undecodable>" }
            return string
        }()

        print("""
        [Auth] ── login succeeded ────────────────────────────────
        [Auth] refresh token returned: \(state.refreshToken != nil ? "YES" : "NO — check that offline_access is granted for this application in authentik")
        [Auth] access token expires:   \(state.lastTokenResponse?.accessTokenExpirationDate.map(String.init(describing:)) ?? "unknown")
        [Auth] scopes granted:         \(state.lastTokenResponse?.scope ?? "none reported")
        [Auth] raw userinfo response:
        \(pretty)
        [Auth] ─────────────────────────────────────────────────────
        """)
        #endif
    }

    // MARK: - Presentation

    private func topViewController() throws -> UIViewController {
        let scenes = UIApplication.shared.connectedScenes.compactMap { $0 as? UIWindowScene }
        let window = scenes.first(where: { $0.activationState == .foregroundActive })?.keyWindow
            ?? scenes.compactMap(\.keyWindow).first

        guard var top = window?.rootViewController else { throw AuthError.noPresenter }
        while let presented = top.presentedViewController { top = presented }
        return top
    }
}

File: android/app/src/main/java/tw/stsa/memberapp/auth/AuthManager.kt
package tw.stsa.memberapp.auth

import android.content.Context
import android.content.Intent
import android.util.Log
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.core.content.edit
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withContext
import net.openid.appauth.AuthState
import net.openid.appauth.AuthorizationException
import net.openid.appauth.AuthorizationRequest
import net.openid.appauth.AuthorizationResponse
import net.openid.appauth.AuthorizationService
import net.openid.appauth.AuthorizationServiceConfiguration
import net.openid.appauth.ResponseTypeValues
import net.openid.appauth.TokenResponse
import org.json.JSONObject
import tw.stsa.memberapp.BuildConfig
import tw.stsa.memberapp.R
import tw.stsa.memberapp.net.HttpResponse
import tw.stsa.memberapp.net.httpGet
import java.time.Instant
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

/** Raised when the sign-in tab was dismissed rather than completed. */
class UserCancelledException : Exception()

/**
 * OIDC authorization-code + PKCE against the STSA authentik tenant.
 *
 * Design is driven by the 5-minute access-token lifetime: nothing ever holds
 * an access token string. [accessToken] goes through
 * `AuthState.performActionWithFreshTokens` on every call, which returns the
 * cached token if it is still valid and silently redeems the refresh token if
 * it is not. Callers get a token that is good *right now* or an error.
 */
class AuthManager(context: Context) {

    private val appContext = context.applicationContext
    private val tokenStore = TokenStore(appContext)
    private val prefs = appContext.getSharedPreferences(PROFILE_PREFS, Context.MODE_PRIVATE)

    /**
     * Held for the app's lifetime rather than per sign-in: it also services
     * every token refresh, and rebuilding it would drop the Custom Tabs service
     * binding that makes the browser hand-off fast.
     */
    private val authService = AuthorizationService(appContext)

    var isLoggedIn by mutableStateOf(false)
        private set
    var profile by mutableStateOf<Profile?>(null)
        private set
    var isBusy by mutableStateOf(false)
        private set

    private var authState: AuthState? = null
    private var serviceConfiguration: AuthorizationServiceConfiguration? = null

    /**
     * What the diagnostics section needs to show, sampled on demand — the token
     * state lives inside AppAuth's object graph and is not observable.
     */
    data class TokenSnapshot(
        val hasRefreshToken: Boolean,
        val accessTokenExpiry: Instant?,
        val scopesGranted: String?,
    )

    init {
        // Synchronous on purpose. Resolving "is there a session" asynchronously
        // would show the welcome screen for a frame and then replace it, which
        // is the exact flash the single-gate rule exists to prevent.
        restore()
    }

    // MARK: - Session lifecycle

    /**
     * Builds the intent that opens the sign-in page, after discovery.
     *
     * The caller launches it with an activity-result launcher and hands the
     * result back to [completeAuthorization]. There is no `resume(url)` step
     * like iOS has: the redirect lands on AppAuth's own receiver activity,
     * which returns it through that result.
     */
    suspend fun authorizationIntent(): Intent {
        isBusy = true
        try {
            val configuration = discoverConfiguration()

            // The standard builder derives a PKCE code_verifier and an S256
            // code_challenge on its own. Do not call setCodeVerifier(null) —
            // that turns PKCE off, and this client is public.
            val request = AuthorizationRequest.Builder(
                configuration,
                AuthConfiguration.CLIENT_ID,
                ResponseTypeValues.CODE,
                AuthConfiguration.REDIRECT_URI,
            ).setScopes(AuthConfiguration.SCOPES).build()

            // AppAuth drives a Custom Tab here, so the sign-in page runs in the
            // browser's session and participates in SSO. Never swap this for a
            // WebView.
            return authService.getAuthorizationRequestIntent(request)
        } catch (error: Throwable) {
            isBusy = false
            throw error
        }
    }

    /** Exchanges the authorization code and loads the profile. */
    suspend fun completeAuthorization(data: Intent?) {
        isBusy = true
        try {
            // A dismissed Custom Tab comes back as RESULT_CANCELED with no data.
            val intent = data ?: throw UserCancelledException()

            val response = AuthorizationResponse.fromIntent(intent)
            val failure = AuthorizationException.fromIntent(intent)
            if (response == null) throw failure ?: authError(R.string.error_token_unavailable)

            val state = AuthState(response, failure)
            val tokens = exchange(response)
            state.update(tokens, null)
            adopt(state)

            val (fetched, rawUserinfo) = fetchProfile()
            profile = fetched
            cache(fetched)

            logLoginResult(rawUserinfo, state)
        } finally {
            isBusy = false
        }
    }

    /** Called when the sign-in launcher returns without ever starting the flow. */
    fun abandonAuthorization() {
        isBusy = false
    }

    /**
     * Drops every local credential.
     *
     * Deliberately local-only. Because the flow runs in the shared browser
     * session, the authentik session cookie survives — the next sign-in may
     * complete without a prompt. If you need to end the IdP session too, that
     * is an RP-initiated logout against the discovery document's
     * `end_session_endpoint`.
     */
    fun logout() {
        val sub = profile?.sub ?: prefs.getString(SUBJECT_KEY, null)
        prefs.edit {
            if (sub != null) remove(profileKey(sub))
            remove(SUBJECT_KEY)
        }
        tokenStore.clear()

        authState = null
        profile = null
        isLoggedIn = false
    }

    // MARK: - Tokens

    /**
     * The only supported way to get an access token.
     *
     * Never store the return value. Access tokens live 5 minutes; call this
     * again for the next request and let AppAuth decide whether a refresh is
     * needed.
     */
    suspend fun accessToken(): String {
        val state = authState ?: throw authError(R.string.error_not_authenticated)

        val token = try {
            suspendCancellableCoroutine<String> { continuation ->
                state.performActionWithFreshTokens(authService) { accessToken, _, error ->
                    when {
                        error != null -> continuation.resumeWithException(error)
                        accessToken != null -> continuation.resume(accessToken)
                        else -> continuation.resumeWithException(
                            authError(R.string.error_token_unavailable)
                        )
                    }
                }
            }
        } catch (error: Throwable) {
            // A session dying is ordinary: authentik rotates refresh tokens, they
            // expire after 30 days, and they can be revoked from the account page.
            // What is not ordinary is what used to happen next — this method threw
            // before it could update [isLoggedIn], so the app stayed on the
            // signed-in screen with a session that could never work again, showing
            // a raw OAuth error and a Retry button that could not succeed. The only
            // way out was to find 登出 in 設定.
            //
            // AppAuth has already folded a permanent OAuth failure into the state,
            // so it can be asked.
            if (state.authorizationException != null) {
                withContext(Dispatchers.Main) { logout() }
                throw authError(R.string.error_session_expired)
            }

            // Transient failure — a rotation may still have landed before it, so
            // keep whatever AppAuth wrote rather than leaving a superseded refresh
            // token in the store.
            persist(state)
            throw error
        }

        // A refresh rotates the token response (and on authentik, often the
        // refresh token too), so write the updated state back.
        persist(state)
        withContext(Dispatchers.Main) { isLoggedIn = state.isAuthorized }

        return token
    }

    /**
     * Builds a request carrying a token that is fresh at the moment of the call.
     * Use this for every outbound API call rather than holding a token.
     */
    suspend fun authorizedGet(url: String): HttpResponse =
        httpGet(url, mapOf("Authorization" to "Bearer ${accessToken()}"))

    /**
     * Renews the access token if it has lapsed, so the next request does not
     * pay for a token exchange first.
     *
     * Called when the app comes to the foreground — not on a timer. An access
     * token is only needed at the moment of a request, and
     * `performActionWithFreshTokens` already renews one transparently, so
     * polling would spend battery and rotate refresh tokens for nothing.
     */
    suspend fun refreshIfNeeded() {
        if (!isLoggedIn) return
        runCatching { accessToken() }

        // The encrypted blob survives an app update but the cached claims can be
        // cleared independently, so a restored session may have no idea who the
        // member is. Re-fetch rather than showing an account screen with blank
        // fields.
        if (profile == null) {
            runCatching { fetchProfile() }.getOrNull()?.let { (fetched, _) ->
                profile = fetched
                cache(fetched)
            }
        }
    }

    fun snapshot(): TokenSnapshot = TokenSnapshot(
        hasRefreshToken = authState?.refreshToken != null,
        accessTokenExpiry = authState?.accessTokenExpirationTime?.let(Instant::ofEpochMilli),
        scopesGranted = authState?.scope,
    )

    // MARK: - Discovery

    private suspend fun discoverConfiguration(): AuthorizationServiceConfiguration {
        serviceConfiguration?.let { return it }

        val configuration = suspendCancellableCoroutine { continuation ->
            AuthorizationServiceConfiguration.fetchFromIssuer(
                AuthConfiguration.ISSUER
            ) { fetched, error ->
                if (fetched != null) {
                    continuation.resume(fetched)
                } else {
                    continuation.resumeWithException(
                        error ?: authError(R.string.error_token_unavailable)
                    )
                }
            }
        }

        serviceConfiguration = configuration
        return configuration
    }

    private suspend fun exchange(response: AuthorizationResponse): TokenResponse =
        suspendCancellableCoroutine { continuation ->
            authService.performTokenRequest(response.createTokenExchangeRequest()) { tokens, error ->
                if (tokens != null) {
                    continuation.resume(tokens)
                } else {
                    continuation.resumeWithException(
                        error ?: authError(R.string.error_token_unavailable)
                    )
                }
            }
        }

    // MARK: - Userinfo

    /**
     * Returns the decoded profile plus the raw body, so the caller can log the
     * exact claim set authentik issued.
     */
    private suspend fun fetchProfile(): Pair<Profile, String> {
        val configuration = discoverConfiguration()
        val endpoint = configuration.discoveryDoc?.userinfoEndpoint
            ?: throw authError(R.string.error_no_userinfo_endpoint)

        val response = authorizedGet(endpoint.toString())
        if (response.status != 200) {
            throw authError(R.string.error_userinfo_failed, response.status, response.body)
        }

        return Profile.decode(response.body) to response.body
    }

    // MARK: - Persistence

    private suspend fun adopt(state: AuthState) {
        authState = state
        isLoggedIn = state.isAuthorized
        persist(state)
    }

    private suspend fun persist(state: AuthState) {
        val payload = state.jsonSerializeString()
        withContext(Dispatchers.IO) { tokenStore.write(payload) }
    }

    private fun restore() {
        val payload = tokenStore.read() ?: return
        val state = runCatching { AuthState.jsonDeserialize(payload) }.getOrNull() ?: return

        authState = state
        isLoggedIn = state.isAuthorized

        val sub = prefs.getString(SUBJECT_KEY, null) ?: return
        val cached = prefs.getString(profileKey(sub), null) ?: return
        profile = runCatching { Profile.decode(cached) }.getOrNull()
    }

    /**
     * Profile claims are not credentials, so they live in plain preferences —
     * but keyed on `sub`, never on email or username.
     */
    private fun cache(profile: Profile) {
        val encoded = runCatching { Profile.json.encodeToString(Profile.serializer(), profile) }
            .getOrNull() ?: return
        prefs.edit {
            putString(profileKey(profile.sub), encoded)
            putString(SUBJECT_KEY, profile.sub)
        }
    }

    // MARK: - Errors

    private fun authError(resource: Int, vararg args: Any): Exception =
        IllegalStateException(appContext.getString(resource, *args))

    // MARK: - Verification logging

    private fun logLoginResult(rawUserinfo: String, state: AuthState) {
        if (!BuildConfig.DEBUG) return

        val pretty = runCatching { JSONObject(rawUserinfo).toString(2) }.getOrDefault(rawUserinfo)
        val refresh = if (state.refreshToken != null) {
            "YES"
        } else {
            "NO — check that offline_access is granted for this application in authentik"
        }
        val expiry = state.accessTokenExpirationTime?.let(Instant::ofEpochMilli)?.toString()
            ?: "unknown"

        Log.d(
            TAG,
            """
            ── login succeeded ────────────────────────────────
            refresh token returned: $refresh
            access token expires:   $expiry
            scopes granted:         ${state.scope ?: "none reported"}
            raw userinfo response:
            $pretty
            ─────────────────────────────────────────────────────
            """.trimIndent(),
        )
    }

    companion object {
        private const val TAG = "Auth"
        private const val PROFILE_PREFS = "profile"
        private const val SUBJECT_KEY = "auth.currentSubject"

        private fun profileKey(sub: String) = "auth.profile.$sub"

        /**
         * True when the person dismissed the sign-in page themselves. Callers
         * should treat this as "nothing happened", not as a failure worth an
         * alert.
         */
        fun isUserCancellation(error: Throwable): Boolean = when (error) {
            is UserCancelledException -> true
            is AuthorizationException ->
                error.type == AuthorizationException.TYPE_GENERAL_ERROR &&
                    error.code == AuthorizationException.GeneralErrors.USER_CANCELED_AUTH_FLOW.code
            else -> false
        }
    }
}
